modules/cas_cache/layer_cache_management.c

```diff
@@ -1346,59 +1345,34 @@ static int _cache_mngt_remove_core_flush(ocf_cache_t cache,
 	if (result < 0)
 		goto unlock;
 
-	core_active = (ocf_core_get_state(core) == ocf_core_state_active);
-
-	if (cmd->detach && !core_active) {
-		printk(KERN_WARNING OCF_PREFIX_SHORT
-				"Cannot detach core which "
-				"is already inactive!\n");
-		return -OCF_ERR_CORE_IN_INACTIVE_STATE;
-	}
-
-	if (core_active) {
-		return _cache_mngt_core_flush_sync(core,
-				true, _cache_read_unlock_put_cmpl);
-	} else if (!ocf_mngt_core_is_dirty(core)) {
+	if (!ocf_mngt_core_is_dirty(core)) {
 		result = 0;
 		goto unlock;
-	} else {
-		printk(KERN_WARNING OCF_PREFIX_SHORT
-				"Cannot remove dirty inactive core "
-				"without force option\n");
-		result = -OCF_ERR_CORE_IN_INACTIVE_STATE;
-		goto unlock;
 	}
 
+	return _cache_mngt_core_flush_sync(core, true,
+				_cache_read_unlock_put_cmpl);
+
 unlock:
 	ocf_mngt_cache_read_unlock(cache);
 put:
 	ocf_mngt_cache_put(cache);
 	return result;
 }
 
-static int _cache_mngt_remove_core_prepare(ocf_cache_t cache, ocf_core_t core,
+static int _cache_mngt_remove_active_core_prepare(ocf_cache_t cache, ocf_core_t core,
 		struct kcas_remove_core *cmd)
 {
 	int result = 0;
-	bool core_active;
 
-	core_active = ocf_core_get_state(core) == ocf_core_state_active;
-
-	if (cmd->detach && !core_active) {
-		printk(KERN_WARNING OCF_PREFIX_SHORT
-				"Cannot detach core which "
-				"is already inactive!\n");
-		return -OCF_ERR_CORE_IN_INACTIVE_STATE;
-	}
+	result = block_dev_destroy_exported_object(core);
+	if (result)
+		return result;
 
-	if (core_active) {
-		result = block_dev_destroy_exported_object(core);
-		if (result)
-			return result;
-	}
+	if (cmd->force_no_flush)
+		return -KCAS_ERR_REMOVED_DIRTY;
```

Comment:
This is an important change (IOCTL handler returning KCAS_ERR_REMOVED_DIRTY for "casadm -R -f"). I would make a separate commit out of this.